content/doc/developer/handling-requests/json.adoc

```diff
@@ -0,0 +1,272 @@
+8. `JsonHttpResponse` is a sub class of `HttpResponse` that indicates that the response content will be JSON
+9. An HTTP status can be set in the response.
+10. The method `get-error500` is added in the example to show how to set a error response as JSON.
+
+
+=== Testing with CURL
+
+A `mvn hpi:run` in the plugin should be enough to run it locally.  Assuming that Jenkins is up at http://localhost:8080/jenkins/ , you should have at this point:
+
+[source,bash]
+----
+curl -XGET \
+    -w "\n STATUS:%{http_code}"  \
+    http://localhost:8080/jenkins/custom-api/get-example-param?paramValue=hello
+
+{"message":"I am Jenkins hello"}
+ STATUS:200
+----
+
+=== Example of test with JenkinsRule
+
+[source,java]
+----
+public class JsonAPITest {
+
+    @Rule
+    public JenkinsRule j = new JenkinsRule();
+
+    private static String GET_API_URL = "custom-api/get-example-param?paramValue=hello";
+
+    @Test
+    public void testGetJSON() throws Exception {
+
+        // Testing a simple GET that should answer 200 OK and a JSON
+        JenkinsRule.JSONWebResponse response = j.getJSON(GET_API_URL);
+        assertTrue(response.getContentAsString().contains("I am JenkinsRule hello"));
+        assertEquals(response.getStatusCode(), 200);
+    }
+
+    @Test
+    public void testAdvancedGetJSON() throws Exception {
+        //Testing a GET that requires the user to be authenticated
+        User admin = User.getById("admin", true);
+        MockAuthorizationStrategy auth = new MockAuthorizationStrategy()
+                .grant(Jenkins.ADMINISTER).everywhere().to(admin);
+
+        j.jenkins.setSecurityRealm(j.createDummySecurityRealm());
+        j.jenkins.setAuthorizationStrategy(auth);
+
+        //We need to setup the webclient
+        //By default if the status code is not ok, WebClient throw an exception
+        //Since we want to assert the error status code, we need to set to false.
+        JenkinsRule.WebClient webClient = j.createWebClient();
+        webClient.setThrowExceptionOnFailingStatusCode(false);
+
+        // - simple call without authentication should be forbidden
+        response = j.getJSON(GET_API_URL, webClient);
+        assertEquals(response.getStatusCode(), 403);
+
+        // - same call but authenticated using withBasicApiToken() should be fine
+        response = j.getJSON(GET_API_URL, webClient.withBasicApiToken(admin));
+        assertEquals(response.getStatusCode(), 200);
+    }
+
+----
+
+== Handling POST with Jenkins
+
+This section shows how to expose an HTTP endpoint that takes a structured JSON Object as input, and do a response with a JSON structured Object.
+For this exmaple the same Object is used as input and output, but you can also use different JSON structure for the response.
+
+Starting from the class `JsonAPI` provided for GET example, add:
+
+[source,java]
+----
+@POST
+@WebMethod(name = "")
```

Comment:
Uh, did you forget to specify the name?

Reply:
I think it's a weird but supported case by [MetaClass.java#L542](https://github.com/jenkinsci/stapler/blob/229a24fab17cb69214622ef442e2be71da0d01cb/core/src/main/java/org/kohsuke/stapler/MetaClass.java#L542) if empty, will use the method name as if the annotation was not used.

Reply:
Hmm, best not to rely on this undocumented behavior? The normal idiom would be

```java
@POST
public JsonHttpResponse doCreate(@JsonBody MyJsonObject body) {…}
```

or if you prefer to be explicit about URLs and dislike Stapler method naming conventions

```java
@POST
@WebMethod(name = "create")
public JsonHttpResponse create(@JsonBody MyJsonObject body) {…}
```

Reply:
thanks, I think I need a following PR to fix this

Reply:
https://github.com/jenkins-infra/jenkins.io/pull/4839